🔗🗿 Combined Representation

mberr opened the pull request:
This PR introduces a representation class for the combination of multiple base representations, and updates `NodePiece` and literal models to use this representations. It also reduces some of the class hierarchy for literal combinations, and updates `Representation` `extra_repr`.

mberr committed:
add combined representation

mberr committed:
update RepresentationTestCase

mberr committed:
fix tests

mberr committed:
implement first working draft

mberr committed:
add todo

cthoyt commented on src/pykeen/nn/representation.py:
```diff
@@ -985,3 +986,84 @@ def _plain_forward(
+        max_id: int,
+        base: OneOrManyHintOrType[Representation] = None,
+        base_kwargs: OneOrManyOptionalKwargs = None,
+        combination: Hint[Callable[[Sequence[torch.FloatTensor]], torch.FloatTensor]] = None,
```
do we have a combination resolver?

cthoyt replied:
e.g., maybe include a simple learnable combination like MLP or some kind of convolution-based combinations

mberr replied:
https://github.com/pykeen/pykeen/pull/964#discussion_r890231638

mberr replied:
for now, I used https://github.com/pykeen/pykeen/pull/964/commits/cebb102530214340396a36e87a072178260932c7 - this may be sub-optimal, since we do not have support for e.g., the reduction necessary in NodePiece where we have two *sequences* of potentially different length of same-sized vectors

mberr replied:
The combination resolver in `pykeen.nn.combinations` is so far limited to combining a pair of representations (and looks a bit over-abstracted to me 😄 ).

mberr commented on src/pykeen/nn/representation.py:
```diff
@@ -985,3 +986,84 @@ def _plain_forward(
+class CombinedRepresentation(Representation):
+    """A combined representation."""
+
+    # TODO: upgrade: NodePiece; Literal
```
there is a comment about this 🙂

mberr committed:
export CombinedRepresentation

mberr committed:
re-use aggregation resolver for now

cthoyt commented on src/pykeen/nn/representation.py:
```diff
@@ -1059,7 +1057,8 @@ def combine(
         """
-        return combination([b._plain_forward(indices=indices) for b in base])
+        # TODO: concat may limit the combination choices
+        return combination(torch.cat([b._plain_forward(indices=indices) for b in base], dim=-1))
```
could the cat([...], dim=-1) in a nn.Module

mberr committed:
add base class for n-ary combination

mberr committed:
add new base class

mberr committed:
add concat + aggregate

mberr committed:
add concat & resolver

mberr committed:
use new combination class

cthoyt commented on src/pykeen/nn/combinations.py:
```diff
@@ -265,3 +339,9 @@ def forward(self, x: torch.FloatTensor, literal: torch.FloatTensor) -> torch.Flo
         return self.dropout(z * h + (1 - z) * x)
+
+
+combination_resolver: ClassResolver[NCombination] = ClassResolver.from_subclasses(
```
these are both "combinations" but very different kinds of things. maybe we need to make more clever names

cthoyt replied:
like how about TensorCombination

mberr replied:
I think they are the same, just that the `NCombination` is a more general case with more than two representations.

mberr replied:
I want to push the combination into the representation module, s.t. you can use any interaction with a literal model without having to explicitly code that.

It would also support e.g. having literals *and* textual representations.

cthoyt commented on src/pykeen/nn/combinations.py:
```diff
@@ -25,6 +26,79 @@
+            the output shape
+        """
+        logger.warning("No symbolic computation of output shape.")
+        return self(xs=[torch.empty(size=shape) for shape in input_shapes]).shape
```
isn't there like a symbolic "meta"-device now for this kind of operation?

mberr replied:
> NotImplementedError: Could not run 'aten::_cat' with arguments from the 'Meta' backend. This could be because the operator doesn't exist for this backend

As soon as you start calculating with it, it will break.

cthoyt replied:
😿

cthoyt replied:
but this implementation is fine as it is too

mberr committed:
update NodePiece to CombinedRepresentations

mberr committed:
update literal models

mberr committed:
update old literal combinations

mberr committed:
rename base class

mberr committed:
remove explicit literal *interaction* test cases

mberr committed:
remove duplicate max_id

mberr committed:
merge classes with exactly one subclass

mberr committed:
fix tests

mberr committed:
remove LiteralInteraction

mberr committed:
fix typing

mberr committed:
remove unused imports

mberr committed:
update README

mberr committed:
annotate literal models with interaction cls

mberr committed:
use annotation for README generation

mberr committed:
update README

mberr committed:
Merge remote-tracking branch 'origin/master' into combined-representation

mberr committed:
use relative import

mberr committed:
use relative import

mberr committed:
remove obsolete code

mberr committed:
add extra_repr to base representation

mberr committed:
move extra repr to CombinedRepresentation

mberr committed:
remove resolved todo

mberr committed:
fix outdated import

mberr committed:
remove unnecessary iter_extra_repr

since combination is a nn.Module, it will appear automatically in the
string; the same holds for base, which is a nn.ModuleList.

mberr committed:
update Representation's iter_extra_repr

mberr committed:
infer unique for combined representations

mberr commented on tests/test_nn/test_representation.py:
```diff
@@ -84,6 +82,8 @@ def _pre_instantiation_hook(self, kwargs: MutableMapping[str, Any]) -> MutableMa
             dims=self.dim,
         )
+        # inferred from triples factory
+        kwargs.pop("max_id")
```
maybe it would be better to allow passing `max_id` and use this ID to check against the inferred one; as this would require touching unrelated code, I would postpone that to another PR

mberr committed:
add test for extra_repr

mberr committed:
fix import order

mberr committed:
fix typing

mberr committed:
reduce code duplication

mberr committed:
infer class name

mberr committed:
add docstring to helper

mberr committed:
add return type

mberr committed:
add return type

mberr committed:
fix typing

mberr committed:
Merge remote-tracking branch 'origin/master' into combined-representation

mberr committed:
remove obsolete code

cthoyt commented on src/pykeen/models/nbase.py:
```diff
@@ -280,11 +280,7 @@ def __init__(
-            Interaction[HeadRepresentation, RelationRepresentation, TailRepresentation],
-            Type[Interaction[HeadRepresentation, RelationRepresentation, TailRepresentation]],
-        ],
+        interaction: HintOrType[Interaction[HeadRepresentation, RelationRepresentation, TailRepresentation]],
```
this isn't quite right, doesn't HintOrType allow for none?

cthoyt replied:
you want `InstOrType` here since this can't be none

cthoyt replied:
well actually `Union[str, InstOrType[...]]` to be the same as the original

mberr replied:
Alternative: set a default interaction

mberr replied:
done in https://github.com/pykeen/pykeen/pull/964/commits/42bf9cebfaa5d62b444bc9e077407432cd41de7b

cthoyt committed:
Various cleanups

cthoyt commented on src/pykeen/models/multimodal/base.py:
```diff
@@ -2,30 +2,43 @@
         self,
         triples_factory: TriplesNumericLiteralsFactory,
-        interaction: LiteralInteraction,
+        interaction: HintOrType[Interaction[torch.FloatTensor, torch.FloatTensor, torch.FloatTensor]],
```
you want `InstOrType` here since this can't be none

mberr replied:
cf. https://github.com/pykeen/pykeen/pull/964#discussion_r892002036

cthoyt commented on src/pykeen/nn/combinations.py:
```diff
@@ -178,26 +174,42 @@ def __init__(
+    def output_shape(self, input_shapes: Sequence[Tuple[int, ...]]) -> Tuple[int, ...]:  # noqa: D102
+        # symbolic output to avoid dtype issue
+        # we only need to consider real part here
+        # FIXME should score_real be real_mod?
```
FIXME.... looks like these aren't being covered by tests

mberr replied:
fixed and covered by tests

cthoyt commented on src/pykeen/nn/node_piece/representations.py:
```diff
@@ -285,39 +283,12 @@ def __init__(
+            max_id=triples_factory.num_entities,
+            base=token_representations,
+            combination=ConcatAggregationCombination,
+            combination_kwargs=dict(aggregation=aggregation, dim=-len(token_representations[0].shape)),
```
maybe still worth saving aggregation index as a variable first to give  it a name

mberr replied:
This is available as the `dim` attribute of the combination.

It is now also included in the output of `repr(instance)`

e.g. for NodePieceRepresentation
```
NodePieceRepresentation(
  max_id=14, shape=(3,), unique=True
  (combination): ConcatAggregationCombination(dim=-2)
)
```

mberr replied:
https://github.com/pykeen/pykeen/pull/964/commits/8844be460cb6b571a4071aee4f22cde5c612afff

cthoyt commented on src/pykeen/nn/representation.py:
```diff
@@ -175,6 +177,21 @@ def reset_parameters(self) -> None:
+        if self.normalizer is not None:
+            yield f"normalizer={self.normalizer}"
+        if self.regularizer is not None:
+            yield f"normalizer={self.regularizer}"
```
typo

mberr replied:
https://github.com/pykeen/pykeen/pull/964/commits/7e2a6ffda4431b864b08666f4baf20cdf2436fc1

regularizer is a `nn.Module`, so we do not need to take care of having that included in the string representation

cthoyt commented on tests/cases.py:
```diff
@@ -2030,18 +2035,21 @@ class NodePieceTestCase(RepresentationTestCase):
     num_triples: ClassVar[int] = 31
 
     def _pre_instantiation_hook(self, kwargs: MutableMapping[str, Any]) -> MutableMapping[str, Any]:  # noqa: D102
+        self.num_entities = self.max_id
```
confusing update

mberr replied:
https://github.com/pykeen/pykeen/pull/964/commits/991cccb72fd113fec7bdd5a0c649e8aa5787facc

cthoyt reviewed (changes requested):
need some tests for concatenation modules now

mberr committed:
add skeleton for combination tests

mberr committed:
rename module

mberr committed:
add literal_shape property

mberr committed:
refactor combinations

mberr committed:
expose via pykeen.nn

mberr committed:
add tests

mberr committed:
implement tests

mberr committed:
fix Representation.iter_extra_repr

mberr committed:
set a default interaction

mberr committed:
rename for consistency

mberr committed:
add todo and remove outdated note

mberr committed:
add dim to extra_repr

mberr committed:
remove unused attribute

mberr committed:
remove shape

shape is now always inferred

mberr committed:
add num_tokens property

mberr committed:
push concat MLP into NodePieceRepresentation

mberr committed:
fix import

mberr committed:
remove duplicate code

mberr committed:
cleanup imports

cthoyt commented on src/pykeen/nn/combination.py:
```diff
@@ -0,0 +1,309 @@
+
+    # docstr-coverage: inherited
+    def forward(self, xs: Sequence[torch.FloatTensor]) -> torch.FloatTensor:  # noqa: D102
+        assert len(xs) == 2
```
is this still necessary?

mberr replied:
yes, because the initialization of gate / combination is hard-coded to be a ConcatProjectionCombination of two

cthoyt commented on src/pykeen/nn/modules.py:
```diff
@@ -2000,7 +1935,7 @@ def latex_visualize(self) -> str:
+    Interaction,
     skip={NormBasedInteraction, FunctionalInteraction, MonotonicAffineTransformationInteraction},
-    suffix=Interaction.__name__,
+    default=TransEInteraction,
```
do we really want a default here?

mberr replied:
what speaks against it?

having a default enables easier instantiation when you just want to play around and do not (yet) care about the exact interaction in-place

The choice is up to discussion, but I took something which operates on single vectors, as a small memory requirement and reasonably well-known.

mberr committed:
Merge remote-tracking branch 'origin/master' into combined-representation

mberr committed:
Merge branch 'master' into combined-representation

mberr committed:
use canonical module

mberr committed:
update README

I am not sure what changed there

mberr committed:
fix missing aggregation_kwargs

mberr committed:
fix text representation tests

mberr committed:
remove obsolete parameter from inductive NodePiece

mberr committed:
fix missing ModuleList around base representations

quite serious error

mberr committed:
remove outdated annotation

mberr committed:
some cleanup on NodePiece aggregation

problem: weights of combination are not part of the module.parameters,
since they are hidden behind the functools.partial

mberr committed:
fix problem with trainable parameters in aggr.

mberr committed:
fix imports

cthoyt committed:
Merge branch 'master' into combined-representation

cthoyt committed:
Update combinations.rst

Trigger CI
